mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs

```diff
@@ -0,0 +1,215 @@
+		// won't cause the program to hang
+		void SetTimeout (int socket)
+		{
+			Timeval timeout = new Timeval { tv_sec = 0, tv_usec = 500000 };
+			if (Syscall.setsockopt (socket, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_RCVTIMEO, timeout) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			if (Syscall.setsockopt (socket, UnixSocketProtocol.SOL_SOCKET, UnixSocketOptionName.SO_SNDTIMEO, timeout) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		void WithSocketPair (Action<int, int> f)
+		{
+			var sockets = new int[2];
+			if (Syscall.socketpair (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0, sockets) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				SetTimeout (sockets[0]);
+				SetTimeout (sockets[1]);
+
+				f (sockets[0], sockets[1]);
+			} finally {
+				if (Syscall.close (sockets[0]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				if (Syscall.close (sockets[1]) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+			}
+		}
+
+		void WithSockets (UnixAddressFamily af, UnixSocketType type, UnixSocketProtocol protocol, Action<int, int> f)
+		{
+			int so1, so2;
+			if ((so1 = Syscall.socket (af, type, protocol)) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			try {
+				if ((so2 = Syscall.socket (af, type, protocol)) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				try {
+					SetTimeout (so1);
+					SetTimeout (so2);
+
+					f (so1, so2);
+				} finally {
+					if (Syscall.close (so2) < 0)
+						UnixMarshal.ThrowExceptionForLastError ();
+				}
+			} finally {
+				if (Syscall.close (so1) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+			}
+		}
+
+		[Test]
+		public void Socket ()
+		{
+			int socket;
+			if ((socket = Syscall.socket (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0)) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+
+			if (Syscall.close (socket) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		[Test]
+		public void SocketPair ()
+		{
+			var sockets = new int[2];
+			if (Syscall.socketpair (UnixAddressFamily.AF_UNIX, UnixSocketType.SOCK_STREAM, 0, sockets) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+
+			if (Syscall.close (sockets[0]) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+			if (Syscall.close (sockets[1]) < 0)
+				UnixMarshal.ThrowExceptionForLastError ();
+		}
+
+		[Test]
+		public void SendRecv ()
+		{
+			WithSocketPair ((so1, so2) => {
+				long ret;
+				var buffer1 = new byte[] { 42, 43, 44 };
+				ret = Syscall.send (so1, buffer1, (ulong) buffer1.Length, 0);
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				var buffer2 = new byte[1024];
+				ret = Syscall.recv (so2, buffer2, (ulong) buffer2.Length, 0);
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				Assert.AreEqual (buffer1.Length, ret);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer2[i]);
+				});
```

Comment:
Indentation on this line should align with the start of `WithSocketPair`:

    WithSocketPair ((so1, so2) => {
        // ...
    });